react/features/calendar-sync/middleware.js

```diff
@@ -0,0 +1,211 @@
+            startDate.getTime(),
+            endDate.getTime(),
+            []
+        )
+        .then(events => {
+            const { knownDomains } = store.getState()['features/calendar-sync'];
+            const eventList = [];
+
+            if (events && events.length) {
+                for (const event of events) {
+                    const jitsiURL = _getURLFromEvent(event, knownDomains);
+                    const now = Date.now();
+
+                    if (jitsiURL) {
+                        const eventStartDate = Date.parse(event.startDate);
+                        const eventEndDate = Date.parse(event.endDate);
+
+                        if (isNaN(eventStartDate) || isNaN(eventEndDate)) {
+                            logger.warn(
+                                'Skipping calendar event due to invalid dates',
+                                event.title,
+                                event.startDate,
+                                event.endDate
+                            );
+                        } else if (eventEndDate > now) {
+                            eventList.push({
+                                endDate: eventEndDate,
+                                id: event.id,
+                                startDate: eventStartDate,
+                                title: event.title,
+                                url: jitsiURL
+                            });
+                        }
+                    }
+                }
+            }
+
+            store.dispatch(updateCalendarEntryList(eventList.sort((a, b) =>
+                a.startDate - b.startDate
+            ).slice(0, MAX_LIST_LENGTH)));
+        })
+        .catch(error => {
+            logger.error('Error fetching calendar.', error);
+        });
+    })
+    .catch(reason => {
+        logger.error('Error accessing calendar.', reason);
+    });
+}
+
+/**
+ * Retreives a jitsi URL from an event if present.
+ *
+ * @private
+ * @param {Object} event - The event to parse.
+ * @param {Array<string>} knownDomains - The known domain names.
+ * @returns {string}
+ *
+ */
+function _getURLFromEvent(event, knownDomains) {
+    const linkTerminatorPattern = '[^\\s<>$]';
+    /* eslint-disable max-len */
+    const urlRegExp
+        = new RegExp(`http(s)?://(${knownDomains.join('|')})/${linkTerminatorPattern}+`, 'gi');
+    /* eslint-enable max-len */
+    const schemeRegExp
+        = new RegExp(`${APP_LINK_SCHEME}${linkTerminatorPattern}+`, 'gi');
+    const fieldsToSearch = [
+        event.title,
+        event.url,
+        event.location,
+        event.notes,
+        event.description
+    ];
+    let matchArray;
+
+    for (const field of fieldsToSearch) {
+        if (typeof field === 'string') {
+            while (
+                (matchArray
+                    = urlRegExp.exec(field) || schemeRegExp.exec(field))
+                        !== null
+            ) {
+                return matchArray[0];
+            }
+        }
+    }
```

Comment:
This seems like a dense chunk of logic to iterate over the fieldsToSearch, maybe just because the indenting for 80 characters is throwing me off. Could you explain what this does?

Reply:
Sure! It goes through all the possible fields (fieldsToSearch array) and tries to match their content to the urlRegExp or the schemeRegExp that looks for a jitsi URL or an app scheme URL respectively. Then it returns the first match.
But now that I reviewed this code I start to feel that the while is not necessary anymore, so thanks! :)